
### SMuRF vignette
by Huang Weitai 

11th Sept 2017

</br>SMuRF is an R package that contains functions for the prediction of a consensus set of somatic mutation calls based on a random forest machine learning classification model. To run these functions, you will need output data from the bcbio-nextgen pipeline http://bcbio-nextgen.readthedocs.io/en/latest/contents/pipelines.html#cancer-variant-calling containing the VCF output for algorithms MuTect2, FreeBayes, VarDict and VarScan. 

In this vignette, we will be using a partial output dataset (https://github.com/skandlab/SMuRF/tree/master/test) derived from the chronic lymphocytic leukemia data downloaded from the European Genome-phenome Archive (EGA) under the accession number EGAS00001001539.

Requirements for package:

Dependencies: https://github.com/skandlab/SMuRF/wiki/Running-SMuRF

<br/>**Installation instructions:**

<br/>1. The latest version of the package is updated on Github https://github.com/skandlab/SMuRF
<br/>2. You can install the current SMuRF directly from Github via the following R command: 
```r
#devtools is required
install.packages("devtools")
library(devtools)
install_github("skandlab/SMuRF", subdir="smurf")
```
(*Alternative option*) SMuRF installation via downloading of the package from Github: 
```r
#Clone or download package from Github https://github.com/skandlab/SMuRF/tree/master/smurf
install.packages("my/current/directory/smurf", repos = NULL, type = "source")
```

Download the test files into your designated directory: https://github.com/skandlab/SMuRF/tree/master/test 
```r
download.file('https://github.com/skandlab/SMuRF/raw/master/test/varscan.vcf.gz','varscan.vcf.gz')
download.file('https://github.com/skandlab/SMuRF/raw/master/test/vardict.vcf.gz','vardict.vcf.gz')
download.file('https://github.com/skandlab/SMuRF/raw/master/test/mutect2.vcf.gz','mutect2.vcf.gz')
download.file('https://github.com/skandlab/SMuRF/raw/master/test/freebayes.vcf.gz','freebayes.vcf.gz')
```

<br/>Before we start using the package's functions, set your designated file directory containing your sample
```r
mydir <- getwd() #get current directory

```
_SMuRF_ allows for the prediction of single somatic nucleotide variants (SNV) as well as small insertions and deletions (indels). In this example we will be predicting both SNVs and INDELs and thus we will be using the "combined" import functionality (other options are "snv" and "indel" only).
```r
library("smurf")
myresults <- smurf(mydir, "combined") #save output into 'myresults' variable
#smurf(mydir, "snv") 
#smurf(mydir, "indel") 
#this will run SMuRF and generate predictions based on input files in 'mydir'

```
_The first time you run SMuRF, required packages may be installed._

Output files saved includes variant statistics (_stats_) and the predicted reads (_predicted_)

```r
# myresults <- smurf(mydir, "combined") 

myresults$smurf_indel$stats_indel

#            Passed_Calls
#Mutect2              383
#FreeBayes             85
#VarDict              149
#VarScan              189
#Atleast1             775
#Atleast2              29
#Atleast3               2
#All4                   0
#SMuRF_INDEL            1

myresults$smurf_indel$predicted_indel

#Chr	START_POS_REF	END_POS_REF	REF	ALT	REF_MFVdVs	ALT_MFVdVs	CALLS	                Sample_Name	Alt_Allele_Freq	SMuRF_score
#1	  17820432	    17820433	  AT	A	  AT/AT/AT/AT	A/A/A/A	    NA/NA/Varscan/Vardict	icgc_cll	  0.495	          0.6210418

myresults$smurf_snv$stats_snv

#          Passed_Calls
#Mutect2           5004
#FreeBayes          115
#VarDict            124
#VarScan            534
#Atleast1          5699
#Atleast2            49
#Atleast3            19
#All4                10
#SMuRF_SNV          399

head(myresults$smurf_snv$predicted_snv)

#Chr	START_POS_REF	END_POS_REF	REF	ALT	REF_MFVdVs	ALT_MFVdVs	CALLS	Sample_Name	Alt_Allele_Freq	SMuRF_score
#1	  1614895	  1614895	G	T	G/NA/NA/NA	T/NA/NA/NA	Mutect2/NA/NA/NA	icgc_cll	0.063	0.4036097
#1	  1945135	  1945135	C	A	C/NA/NA/NA	A/NA/NA/NA	Mutect2/NA/NA/NA	icgc_cll	0.037	0.4314701
#1	  2119800	  2119800	C	A	C/NA/NA/NA	A/NA/NA/NA	Mutect2/NA/NA/NA	icgc_cll	0.052	0.4926129
#1	  2180985	  2180985	A	G	A/A/A/A	G/G/G/G	Mutect2/Freebayes/Varscan/Vardict	icgc_cll	0.479	0.9950507
#1	  2762363	  2762363	G	T	G/NA/NA/NA	T/NA/NA/NA	Mutect2/NA/NA/NA	icgc_cll	0.053	0.4889364


```
**Output file description/legend**

Column Name | Description
----------- | -------------------------------------------------------------------------------------------------
Chr         | Chromosome number
START_POS_REF/END_POS_REF         | Start and End nucleotide position of the somatic mutation
REF/ALT     | Consensus Ref and Alt nucleotide changes of the highest likelihood
REF_MFVdVs/ALT_MFVdVs        | Reference and Alternative nucleotide changes from each caller; Mutect2 (M), Freebayes (F), Vardict (Vd), Varscan (Vs)
CALLS       | List of callers which the mutation is detected
Sample_Name | Sample name is extracted based on your labeled samples in the vcf files
Alt_Allele_Freq | Mean Variant allele frequency calculated from the tumor reads of the callers
SMuRF_score      | SMuRF confidence score of the predicted mutation


</br>You may also retrieve the time taken for your run.
```r
myresults$time.taken

#Time difference of 20.52405 secs
```

You can check the parsed output used for the prediction:
```r
myresults$smurf_indel$parse_indel

myresults$smurf_snv$parse_snv

```

You may check the output files generated by the test samples in this section to the expected results we provided located in the _results_ folder https://github.com/skandlab/SMuRF/tree/master/test/results.



<br/>**Running on multiple samples**

Use our R package to efficiently do somatic mutation predictions on multiple matched tumor-normal samples by providing the list of directories of where your sample files are located. 
```r
#Example

sample_directories <- list("my/dir/sample_A", "my/dir/sample_B", "my/dir/sample_C")

myresults <- list()

for(i in 1:length(sample_directories))
 {
 myresults[[i]] <- smurf(sample_directories[i], "combined")
 } 
 
#myresults[[1]]$time.taken
#Time difference of 9.973997 secs

#myresults[[2]]$time.taken
#Time difference of 11.1712 secs

#myresults[[3]]$time.taken
#Time difference of 15.18325 secs
```

For errors and bugs, 
please report on our Github page.

  
  
  
  
  
  
  

