---
title: "SMuRF vignette"
author: "Huang Weitai"
date: "9 June 2017"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{My First Vignette}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

SMuRF is an R package that contains functions for the prediction of a consensus set of somatic mutation calls based on a random forest machine learning classification model. To run these functions, you will need output data from the bcbio-nextgen pipeline http://bcbio-nextgen.readthedocs.io/en/latest/contents/pipelines.html#cancer-variant-calling containing the VCF output for algorithms MuTect2, FreeBayes, VarDict and VarScan. 

In this vignette, we will be using a partial output dataset (https://github.com/skandlab/SMuRF/tree/master/test) derived from the chronic lymphocytic leukemia data downloaded from the European Genome-phenome Archive (EGA) under the accession number EGAS00001001539.

Requirements for package:

Dependencies: https://github.com/skandlab/SMuRF/wiki/Running-SMuRF

<br/>**Installation instructions:**

<br/>1. The latest version of the package is updated on Github https://github.com/skandlab/SMuRF
<br/>2. You can install the current SMuRF directly from Github via the following R command: 
```r
#devtools is required
install.packages("devtools")
library(devtools)
install_github("skandlab/SMuRF", subdir="smurf")
```
(*Alternative option*) SMuRF installation via downloading of the package from Github: 
```r
#Clone or download package from Github https://github.com/skandlab/SMuRF/tree/master/smurf
install.packages("my/current/directory/smurf", repos = NULL, type = "source")
```

Download the test files into your designated directory: https://github.com/skandlab/SMuRF/tree/master/test 
```r
download.file('https://github.com/skandlab/SMuRF/raw/master/test/varscan.vcf.gz','varscan.vcf.gz')
download.file('https://github.com/skandlab/SMuRF/raw/master/test/vardict.vcf.gz','vardict.vcf.gz')
download.file('https://github.com/skandlab/SMuRF/raw/master/test/mutect2.vcf.gz','mutect2.vcf.gz')
download.file('https://github.com/skandlab/SMuRF/raw/master/test/freebayes.vcf.gz','freebayes.vcf.gz')
```

<br/>Before we start using the package's functions, set your designated file directory containing your sample
```r
mydir <- getwd() #get current directory

```
_SMuRF_ allows for the prediction of single somatic nucleotide variants (SNV) as well as small insertions and deletions (indels). In this example we will be predicting both SNVs and INDELs and thus we will be using the "combined" import functionality (other options are "snv" and "indel" only).
```r
library("smurf")
smurf(mydir, "combined") 
#smurf(mydir, "snv") 
#smurf(mydir, "indel") 
#this will run SMuRF and generate predictions based on input files in 'mydir'

```
_The first time you run SMuRF, required packages may be installed._

Output files saved includes variant statistics (_stats_) and the predicted reads (_predicted_)

```r
# myresults <- smurf(mydir, "combined") 

myresults$smurf_indel$stats_indel

#            Passed_Calls
#Mutect2              383
#FreeBayes             85
#VarDict              166
#VarScan              158
#Atleast1             765
#Atleast2              25
#Atleast3               2
#All4                   0
#SMuRF_INDEL            1

myresults$smurf_indel$predicted_indel

#Chr START_POS_REF END_POS_REF REF_MFVdVs    ALT_MFVdVs SMuRF_score
#1 "1" "17820432"    "17820433"  "AT/AT/AT/AT" "A/A/A/A"  "0.571875" 

myresults$smurf_snv$stats_snv

myresults$smurf_snv$predict_snv

```
**Output file description/legend**

Column Name | Description
----------- | -------------------------------------------------------------------------------------------------
Chr         | Chromosome number
START_POS_REF/END_POS_REF         | Start and End nucleotide position of the somatic mutation
REF_MFVdVs/ALT_MFVdVs        | Reference and Alternative nucleotide changes from each caller; Mutect2 (M), Freebayes (F), Vardict (Vd), Varscan (Vs)
SMuRF_score      | SMuRF confidence score of the predicted mutation


</br>You may also retrieve the time taken for your run.
```r
myresults$time.taken

#Time difference of 20.52405 secs
```

You can check the parsed output used for the prediction:
```r
myresults$smurf_indel$parse_indel

myresults$smurf_snv$parse_snv

```

You may check the output files generated by the test samples in this section to the expected results we provided located in the _results_ folder https://github.com/skandlab/SMuRF/tree/master/test/results.



<br/>**Running on multiple samples**

Use our R package to efficiently do somatic mutation predictions on multiple matched tumor-normal samples by providing the list of directories of where your sample files are located. 
```r
#Example

sample_directories <- list("my/dir/sample_A", "my/dir/sample_B", "my/dir/sample_C")

myresults <- list()

for(i in 1:length(sample_directories))
 {
 myresults[[i]] <- smurf(sample_directories[i], "combined")
 } 
 
#myresults[[1]]$time.taken
#Time difference of 9.973997 secs

#myresults[[2]]$time.taken
#Time difference of 11.1712 secs

#myresults[[3]]$time.taken
#Time difference of 15.18325 secs
```

For errors and bugs, 
please report on our Github page.

  
  
  
  
  
  
  

